pybop/plotting/plot2d.py

```diff
@@ -70,6 +70,21 @@ def plot2d(
         for j, yj in enumerate(y):
             costs[j, i] = cost(np.array([xi, yj]))
 
+    if plot_optim:
```

Comment:
Is the aim of this codeblock to append the locations the optimiser visited to the cost landscapes? At the moment, it breaks the gradient plots below it due to `cost` function being overwritten as a `np.array`, and as such isn't used in the plot2d object.

I added the below to line 127 of `unit/test_plots.py` to catch the error.
```python
# Plot gradient cost landscape
pybop.plot2d(optim, gradient=True, steps=5)
```

Reply:
Great catch! Yes, it should have been `costs` not `cost`... and thanks for the test.

From further testing, I've seen that the extra points (and nans) can modify the contour plot in unexpected ways. So I have implemented interpolation to avoid the nans and made this feature optional with the switch `use_optim_log`.